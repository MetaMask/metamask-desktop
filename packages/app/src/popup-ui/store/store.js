import { createStore, applyMiddleware } from 'redux';
import thunkMiddleware from 'redux-thunk';
import rootReducer from '../ducks';

export default function configureStore(initialState = {}) {
  const storeEnhancers = applyMiddleware(thunkMiddleware);
  const store = createStore(rootReducer, initialState, storeEnhancers);
  return { store };
}
