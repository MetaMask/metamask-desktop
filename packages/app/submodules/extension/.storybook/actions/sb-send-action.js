export const updateSendAsset = (type) => ({
  type: 'send/updateSendAsset',
  payload: type,
});

export const updateSendStage = (stage) => ({
  type: 'send/updateSendStage',
  payload: stage,
});
