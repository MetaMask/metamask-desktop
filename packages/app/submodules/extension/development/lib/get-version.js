const { version: manifestVersion } = require('../../package.json');
const { BuildType } = require('./build-type');

/**
 * Get the current version of the MetaMask extension. The base manifest version
 * is modified according to the build type and version.
 *
 * The build version is needed because certain build types (such as beta) may
 * be released multiple times during the release process.
 *
 * @param {BuildType} buildType - The build type.
 * @param {number} buildVersion - The build version.
 * @returns {string} The MetaMask extension version.
 */
function getVersion(buildType, buildVersion) {
  return buildType === BuildType.main
    ? manifestVersion
    : `${manifestVersion}-${buildType}.${buildVersion}`;
}

module.exports = { getVersion };
