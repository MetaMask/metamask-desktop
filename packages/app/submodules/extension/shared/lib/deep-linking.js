export function openCustomProtocol(protocolLink) {
  return new Promise((resolve, reject) => {
    if (window?.navigator?.msLaunchUri) {
      window.navigator.msLaunchUri(protocolLink, resolve, () => {
        reject(new Error('Failed to open custom protocol link'));
      });
    } else {
      const timeoutId = window.setTimeout(function () {
        reject(new Error('Timeout opening custom protocol link'));
      }, 500);
      window.addEventListener('blur', function () {
        window.clearTimeout(timeoutId);
        resolve();
      });
      window.location = protocolLink;
    }
  });
}
