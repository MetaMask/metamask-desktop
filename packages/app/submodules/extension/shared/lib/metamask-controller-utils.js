import { conversionUtil } from '../modules/conversion.utils';

export function hexToDecimal(hexValue) {
  return conversionUtil(hexValue, {
    fromNumericBase: 'hex',
    toNumericBase: 'dec',
  });
}

export function getTokenValueParam(tokenData = {}) {
  return tokenData?.args?._value?.toString();
}
