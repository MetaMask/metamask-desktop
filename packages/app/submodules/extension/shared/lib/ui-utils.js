let _supportLink = 'https://support.metamask.io';

///: BEGIN:ONLY_INCLUDE_IN(flask,desktopextension,desktopapp)
_supportLink = 'https://metamask-flask.zendesk.com/hc';
///: END:ONLY_INCLUDE_IN

export const SUPPORT_LINK = _supportLink;
