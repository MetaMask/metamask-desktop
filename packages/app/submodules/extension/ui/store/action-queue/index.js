import pify from 'pify';
import { isManifestV3 } from '../../../shared/modules/mv3.utils';

// // A simplified pify maybe?
// function pify(apiObject) {
//   return Object.keys(apiObject).reduce((promisifiedAPI, key) => {
//     if (apiObject[key].apply) { // depending on our browser support we might use a nicer check for functions here
//       promisifiedAPI[key] = function (...args) {
//         return new Promise((resolve, reject) => {
//           return apiObject[key](
//             ...args,
//             (err, result) => {
//               if (err) {
//                 reject(err);
//               } else {
//                 resolve(result);
//               }
//             },
//           );
//         });
//       };
//     }
//     return promisifiedAPI;
//   }, {});
// }

let background = null;
let promisifiedBackground = null;

const actionRetryQueue = [];

export const generateActionId = () => Date.now() + Math.random();

function failQueue() {
  actionRetryQueue.forEach(({ reject }) =>
    reject(
      Error('Background operation cancelled while waiting for connection.'),
    ),
  );
}

/**
 * Drops the entire actions queue. Rejects all actions in the queue unless silently==true
 * Does not affect the single action that is currently being processed.
 *
 * @param {boolean} [silently]
 */
export function dropQueue(silently) {
  if (!silently) {
    failQueue();
  }
  actionRetryQueue.length = 0;
}

// add action to queue
const executeActionOrAddToRetryQueue = (item) => {
  if (actionRetryQueue.some((act) => act.actionId === item.actionId)) {
    return;
  }

  if (background.connectionStream.readable) {
    executeAction({
      action: item,
      disconnectSideeffect: () => actionRetryQueue.push(item),
    });
  } else {
    actionRetryQueue.push(item);
  }
};

/**
 * Promise-style call to background method
 * In MV2: invokes promisifiedBackground method directly.
 * In MV3: action is added to retry queue, along with resolve handler to be executed on completion,
 *  the queue is then immediately processed if background connection is available.
 *  On completion (successful or error) the action is removed from the retry queue.
 *
 * @param {string} method - name of the background method
 * @param {Array} [args] - arguments to that method, if any
 * @param {any} [actionId] - if an action with the === same id is submitted, it'll be ignored if already in queue waiting for a retry.
 * @returns {Promise}
 */
export function submitRequestToBackground(
  method,
  args = [],
  actionId = generateActionId(), // current date is not guaranteed to be unique
) {
  if (isManifestV3) {
    return new Promise((resolve, reject) => {
      executeActionOrAddToRetryQueue({
        actionId,
        request: { method, args },
        resolve,
        reject,
      });
    });
  }
  return promisifiedBackground[method](...args);
}

/**
 * Callback-style call to background method
 * In MV2: invokes promisifiedBackground method directly.
 * In MV3: action is added to retry queue, along with resolve handler to be executed on completion,
 *  the queue is then immediately processed if background connection is available.
 *  On completion (successful or error) the action is removed from the retry queue.
 *
 * @param {string} method - name of the background method
 * @param {Array} [args] - arguments to that method, if any
 * @param callback - Node style (error, result) callback for finishing the operation
 * @param {any} [actionId] - if an action with the === same id is submitted, it'll be ignored if already in queue.
 */
export const callBackgroundMethod = (
  method,
  args = [],
  callback,
  actionId = generateActionId(), // current date is not guaranteed to be unique
) => {
  if (isManifestV3) {
    const resolve = (value) => callback(null, value);
    const reject = (err) => callback(err);
    executeActionOrAddToRetryQueue({
      actionId,
      request: { method, args },
      resolve,
      reject,
    });
  } else {
    background[method](...args, callback);
  }
};

async function executeAction({ action, disconnectSideeffect }) {
  const {
    request: { method, args },
    resolve,
    reject,
  } = action;
  try {
    resolve(await promisifiedBackground[method](...args));
  } catch (err) {
    if (
      background.DisconnectError && // necessary to not break compatibility with background stubs or non-default implementations
      err instanceof background.DisconnectError
    ) {
      disconnectSideeffect(action);
    } else {
      reject(err);
    }
  }
}

let processingQueue = false;

// Clears list of pending action in actionRetryQueue
// The results of background calls are wired up to the original promises that's been returned
// The first method on the queue gets called synchronously to make testing and reasoning about
//  a single request to an open connection easier.
async function processActionRetryQueue() {
  if (processingQueue) {
    return;
  }
  processingQueue = true;
  try {
    while (
      background.connectionStream.readable &&
      actionRetryQueue.length > 0
    ) {
      // If background disconnects and fails the action, the next one will not be taken off the queue.
      // Retrying an action that failed because of connection loss while it was processing is not supported.
      const item = actionRetryQueue.shift();
      await executeAction({
        action: item,
        disconnectSideeffect: () => actionRetryQueue.unshift(item),
      });
    }
  } catch (e) {
    // error in the queue mechanism itself, the action was malformed
    console.error(e);
  }
  processingQueue = false;
}

/**
 * Sets/replaces the background connection reference
 * Under MV3 it also triggers queue processing if the new background is connected
 *
 * @param {*} backgroundConnection
 */
export async function _setBackgroundConnection(backgroundConnection) {
  background = backgroundConnection;
  promisifiedBackground = pify(background);
  if (isManifestV3) {
    if (processingQueue) {
      console.warn(
        '_setBackgroundConnection called while a queue was processing and not disconnected yet',
      );
    }
    // Process all actions collected while connection stream was not available.
    processActionRetryQueue();
  }
}
