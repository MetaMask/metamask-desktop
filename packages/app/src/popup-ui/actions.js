import {
  setCurrentLocale,
  updateMetamaskState,
} from '../../submodules/extension/ui/store/actions';
import * as actionConstants from '../../submodules/extension/ui/store/actionConstants';
import { fetchLocale } from '../../submodules/extension/ui/helpers/utils/i18n-helper';

export function updateState(newState) {
  return async (dispatch, getState) => {
    const isFirstState = !getState().metamask.provider;

    if (isFirstState) {
      dispatch({
        type: actionConstants.UPDATE_METAMASK_STATE,
        value: newState,
      });

      const locale = newState.currentLocale;
      const localeMessages = await fetchLocale(locale);

      dispatch(setCurrentLocale(locale, localeMessages));
    } else {
      updateMetamaskState(newState)(dispatch, getState);
    }
  };
}
