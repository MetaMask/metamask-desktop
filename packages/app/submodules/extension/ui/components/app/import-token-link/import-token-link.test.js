import React from 'react';
import configureMockStore from 'redux-mock-store';
import { fireEvent, screen } from '@testing-library/react';
import { detectNewTokens } from '../../../store/actions';
import { renderWithProvider } from '../../../../test/lib/render-helpers';
import ImportTokenLink from '.';

const mockPushHistory = jest.fn();

jest.mock('react-router-dom', () => {
  const original = jest.requireActual('react-router-dom');
  return {
    ...original,
    useLocation: jest.fn(() => ({ search: '' })),
    useHistory: () => ({
      push: mockPushHistory,
    }),
  };
});

jest.mock('../../../store/actions.js', () => ({
  detectNewTokens: jest.fn(),
}));

describe('Import Token Link', () => {
  it('should match snapshot for goerli chainId', () => {
    const mockState = {
      metamask: {
        provider: {
          chainId: '0x5',
        },
      },
    };

    const store = configureMockStore()(mockState);

    const { container } = renderWithProvider(<ImportTokenLink />, store);

    expect(container).toMatchSnapshot();
  });

  it('should match snapshot for mainnet chainId', () => {
    const mockState = {
      metamask: {
        provider: {
          chainId: '0x1',
        },
      },
    };

    const store = configureMockStore()(mockState);

    const { container } = renderWithProvider(<ImportTokenLink />, store);

    expect(container).toMatchSnapshot();
  });

  it('should detectNewTokens when clicking refresh', () => {
    const mockState = {
      metamask: {
        provider: {
          chainId: '0x5',
        },
      },
    };

    const store = configureMockStore()(mockState);

    renderWithProvider(<ImportTokenLink />, store);

    const refreshList = screen.getByTestId('refresh-list-button');
    fireEvent.click(refreshList);

    expect(detectNewTokens).toHaveBeenCalled();
  });

  it('should push import token route', () => {
    const mockState = {
      metamask: {
        provider: {
          chainId: '0x5',
        },
      },
    };

    const store = configureMockStore()(mockState);

    renderWithProvider(<ImportTokenLink />, store);

    const importToken = screen.getByTestId('import-token-button');
    fireEvent.click(importToken);

    expect(mockPushHistory).toHaveBeenCalledWith('/import-token');
  });
});
