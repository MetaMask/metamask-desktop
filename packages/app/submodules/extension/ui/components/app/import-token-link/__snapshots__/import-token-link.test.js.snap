// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Import Token Link should match snapshot for goerli chainId 1`] = `
<div>
  <div
    class="box import-token-link box--flex-direction-row box--text-align-center"
  >
    <a
      class="button btn-link import-token-link__link"
      data-testid="refresh-list-button"
      role="button"
      tabindex="0"
    >
      Refresh list
    </a>
    or
    <a
      class="button btn-link import-token-link__link"
      data-testid="import-token-button"
      role="button"
      tabindex="0"
    >
      import tokens
    </a>
  </div>
</div>
`;

exports[`Import Token Link should match snapshot for mainnet chainId 1`] = `
<div>
  <div
    class="box import-token-link box--flex-direction-row box--text-align-center"
  >
    <a
      class="button btn-link import-token-link__link"
      data-testid="refresh-list-button"
      role="button"
      tabindex="0"
    >
      Refresh list
    </a>
    or
    <a
      class="button btn-link import-token-link__link"
      data-testid="import-token-button"
      role="button"
      tabindex="0"
    >
      import tokens
    </a>
  </div>
</div>
`;
